### DISCUSSION TITLE
No reboot after 2.1.0 uninstallation

### DISCUSSION THREAD

**Discussion by dswiss-desktop-app** [dokan-dev/dokany#1208] [Q&A]
Hi, we've noticed that uninstalling Dokan 2.1.0 does not ask for a reboot anymore - it still needs to be done manually to install it again, or to install a different version. Is it intentional ? To me it may lead to problems with upgrades in future.
Cheers

**Comment by Liryna** [comment]
Normally it should ask for restart if the installer wasn't able to remove the driver during uninstall. Usually it happens when the driver was used (mount a filesystem).

**Reply by massimo-marino** [reply]
Hi @Liryna ,
I confirm this is an issue for dokan 2.1.0 (and 2.2.0 too). Reboot is not triggered when uninstalled.
I made some investigation on this issue and this is what I found.

**My system specs:**
Edition	Windows 11 Pro
Version		23H2
Installed on	‎17/‎10/‎2022
OS build	22631.4037
Experience	Windows Feature Experience Pack 1000.22700.1027.0

Device name	LAPTOP-LNSL1504
Processor	11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz   2.42 GHz
Installed RAM	16.0 GB (15.8 GB usable)
Device ID	47046800-1EE3-43E7-A08E-629186209EB7
Product ID	00330-54004-39143-AAOEM
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display

I used Dokan_x64.msi for 2.0.4 and 2.1.0.
I followed these steps being sure the registry does not contain any entries, and no leftovers  from previous installations:

- run Dokan_x64.msi to install dokan
- immediately after, run it again to uninstall dokan

For 2.0.4: the restart is always prompted when the uninstaller ends. Here is the screenshot:
![dokan-2 0 4-uninst-ends-with-restart](https://github.com/user-attachments/assets/37031be1-01d4-403a-acf6-c2076a85518f)

For 2.1.0 it never occur.

The registry for 2.0.4 at the end of the uninstallation (before restart):
![dokan-2 0 4-uninstall-registry](https://github.com/user-attachments/assets/bf6cd916-033f-4c37-b0ab-219274af32c1)
![dokan-2 0 4-uninstall-registry-2](https://github.com/user-attachments/assets/31d55d85-f9a4-4a37-bf33-88a91ccbb40f)

On, the other end, once the uninstaller for 2.1.0 is ended (and no restart is prompted), if I try again to immediately install 2.1.0 it fails because a Windows reboot is still pending. These are the screenshots with the registry contents once the installer has failed:
![dokan-2 1 0-inst-failed](https://github.com/user-attachments/assets/1dbddd7f-2424-4840-b4c8-69997879cad2)
![dokan-2 1 0-inst-failed-2](https://github.com/user-attachments/assets/c31d5408-19da-457a-99a2-7f8647625456)

See these links about why HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\ seems to be important in this context:

https://community.cryptomator.org/t/dokan-aktualisierung-fehler-0x80070643-bei-der-installation-deinstallation/6494
https://learn.microsoft.com/en-us/previous-versions/office/exchange-server-analyzer/cc164360(v=exchg.80)

For me this is definitely something to be investigated further.
Regards.

**Comment by dswiss-desktop-app** [comment]
Hi Liryna,

Thanks for the quick reply. Unfortunately that's not what we are experiencing. On all our Windows 11 machines (including a brand new virtual machine where there was no previous Dokan installation) the 2.1.0 uninstallation (we use the DokanSetup.exe bundle) does not trigger the restart, nor the pop-up asking for a restart. After that when we try to install any Dokan version, we get an error saying that there is a reboot pending. I am attaching the 2.1.0 uninstallation log followed by a failed 2.1.0 installation attempt. I have also attached a 2.0.4 uninstaller log where it asks for a restart at the end of the uninstallation.
[Dokan_Library_2.1.0.1000_Bundle_20240418105841_000_Dokan_x64.msi.log](https://github.com/dokan-dev/dokany/files/15022611/Dokan_Library_2.1.0.1000_Bundle_20240418105841_000_Dokan_x64.msi.log)
[Dokan_Library_2.1.0.1000_Bundle_20240418105902_000_Dokan_x64.msi.log](https://github.com/dokan-dev/dokany/files/15022613/Dokan_Library_2.1.0.1000_Bundle_20240418105902_000_Dokan_x64.msi.log)
[Dokan_Library_2.0.4.1000_Bundle_20240418110921_000_Dokan_x64.msi.log](https://github.com/dokan-dev/dokany/files/15022614/Dokan_Library_2.0.4.1000_Bundle_20240418110921_000_Dokan_x64.msi.log)

**Reply by Liryna** [reply]
CC @massimo-marino 

The reboot was indeed gone when we switched to using `dokanctl` in the custom action to install the driver (previously we used wix api to install the driver with the `inf` but we now do it manually (dokanctl)).
I added a custom action that will detect whether reboot is needed. Does that looks alright ?
https://github.com/dokan-dev/dokany/commit/2084b3c872538a3357a21b7ce04b27a39a21f527

![image](https://github.com/user-attachments/assets/b0cddb1c-bac4-4aa1-92d6-768b5d392d89)

**Reply by massimo-marino** [reply]
@Liryna  Yes, it looks good.
